mailet/standard/src/main/java/org/apache/james/transport/matchers/HostIsLocal.java

```diff
@@ -21,17 +21,31 @@
 
 package org.apache.james.transport.matchers;
 
+import java.util.Collection;
+
+import javax.mail.MessagingException;
+
 import org.apache.james.core.MailAddress;
-import org.apache.mailet.base.GenericRecipientMatcher;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMatcher;
+
+import com.github.steveash.guavate.Guavate;
 
 /**
  * Matches mail to Domains which are local
  * .
  */
-public class HostIsLocal extends GenericRecipientMatcher {
-
+public class HostIsLocal extends GenericMatcher {
     @Override
-    public boolean matchRecipient(MailAddress recipient) {
-        return getMailetContext().isLocalServer(recipient.getDomain());
+    public Collection<MailAddress> match(Mail mail) throws MessagingException {
+        return mail.getRecipients()
```

Comment:
is it the code replacing
```java
    @Override
    public final Collection<MailAddress> match(Mail mail) throws MessagingException {
        Collection<MailAddress> matching = new ArrayList<>();
        for (MailAddress recipient : mail.getRecipients()) {
            if (matchRecipient(recipient)) {
                matching.add(recipient);
            }
        }
        return ImmutableList.copyOf(matching);
    }
```

If so, I don't see how it is an improvement

Reply:
Well grouping calls for domain is an improvment, we loose on the readability side that's clear.

However I expect from a james bundled matcher to not be obviously inneficient.

Reply:
I don't disagree about the optimization, I was just unable to understand it was doing something different, sorry

Reply:
now that i parsed it successfully, I suggest:
```
recipientsByDomain(mail)
.flatMap(localUsers)
.collect(Guavate.toImmutableList());
```